### ISSUE TITLE
`/` may be better handled by ex-mode

### ISSUE THREAD

**Issue by lloeki** [atom/vim-mode#596]
`/` is in fact a shortcut to `:/`. From vim's `:help :`:

> Command-line mode is used to enter Ex commands (":"), search patterns ("/" and "?"), and filter commands ("!").

Besides, this may help vim-mode get rid of some code such as the search view, which is duplicated (probably due to aforementioned design) between vim-mode and ex-mode.

This may help fix issues such as #578.

**Comment by jazzpi** [comment]
I'm not quite sure about this. Since <kbd>/</kbd> is closely related to <kbd>n</kbd>, <kbd>N</kbd>, <kbd>*</kbd> and <kbd>#</kbd>, those should then probably also be handled by ex-mode as well as potential highlighting...
See also #458 and #493.
As well as that, forcing users to install `ex-mode` as a whole just to be able to use <kbd>/</kbd> seems a bit overkill.

**Comment by dimpase** [comment]
Historically, vi(m) is an add-on of ex, and as such it's only logical that ex-mode and vim-mode should be basically one mode. Vim-mode without ex-mode feels rather feature-poor for any advanced vi(m) user.

**Comment by jazzpi** [comment]
Sure, but a lot of people are apparently fine with vim-mode without ex-mode (vim-mode has about 40x as many downloads as ex-mode). See #50 about the reasoning behind splitting ex into a different package. Following that philosophy, I don't think putting <kbd>/</kbd> into ex-mode is a good idea, although it would simplify some things.

**Comment by dimpase** [comment]
arguments in #50 are just childish; after 20+ years of using vi(m) I don't want to waste time learning yet another way to do complicated things like formatting paragraphs and complicated regex-based replacements, really, and surely I'm not alone.

And measuring success by number of downloads makes little sense, either. You know, pr0n downloads surely beat Atom downloads by many orders of magnitude. ;-)

**Comment by nnjpp** [comment]
I view ex-mode's download count as a sign of a problem. I suspect that it has less downloads because not enough people are willing to dig into github issues to find out that it even exists. I would bet that abandonment comes first in many cases. I'm not buying the notion that only 2.5% of vim users value ex commands.

**Comment by jacekkopecky** [comment]
This is a great discussion to have.

Ex is old and powerful. I don't know that anybody has done an ergonomic comparison between ex-mode and atom ways of doing things that don't fit within the few keystrokes of vim-mode, and I dare not guess which way the comparison would favour. I for one haven't missed ex-mode yet.

The reasoning for having ex-mode outside of vim-mode is the size of the resulting package, correct? 

Would anybody be willing to measure the numbers? How much bigger would the memory footprint of a combined package be, in comparison to having just vim-mode installed, and also in comparison to having both? How much longer would Atom startup take in those three situations?

While I haven't missed ex-mode yet, I don't think I would mind having it. I might even choose to disable it, just to see whether using more of the keyboard than ex-mode does would keep working well for me. But I suspect I wouldn't perceive the added bulk if ex-mode was integrated in vim-mode.

@nnjpp raises a valid point, IMO. Please consider this a vote for bringing ex-mode into vim-mode.

**Comment by bhuga** [comment]
> / is in fact a shortcut to :/. From vim's :help ::

Perhaps `/` is an alias to an ex command, but vim doesn't treat it that way. `/foo` composes as a motion. `c/foo` and `yn` demonstrate how it's not just an ex command. `/` is and was correct to treat as vim instead of ex. Compare to `:r !ls`, which vim does not even repeat with `.`. It's a fundamentally different thing than a command-mode action. 

> The reasoning for having ex-mode outside of vim-mode is the size of the resulting package, correct?

This was not the reason I decided to draw a line between vim and ex in https://github.com/atom/vim-mode/issues/50.

Because vim treats them so differently, commands like `:r !ls` are a fundamentally different from a code perspective, not a memory and footprint perspective. After pressing `:`, a whole new language is parsed and evaluated, one that ignores and is completely separate from the vim motion/command composition stack. I made the initial decision to keep *concerns separated* and *code clean.* Smaller, focused packages are idiomatic in Atom.

I'm not the maintainer anymore, and perhaps the community is ready to try something else. But that's the historical color.

**Comment by ardrigh** [comment]
I have to agree with @nnjpp  comment on the abandonware. If someone reviewed the comments around the vim mode and ex-mode and the long term issues, comments regarding going back to vim or giving up are there to see.

From someone who is learning the workarounds of how vim-mode and ex-mode and Atom-mode collide, it's very jarring bad user experience to the point I think a lot of long-time Vim users will give up and try something else.

I know I have thought about it a few times, but I am trying to get happy with using Atom.

**Comment by jazzpi** [comment]
After thinking about this again, I noticed that `vim` doesn't only treat <kbd>/</kbd> like a motion, it treats *all* `ex` commands like a motion.

So I vote to either 
 - merge `ex-` and `vim-mode` or
 - expose motions via a service and put the link to `ex-mode` in a more prominent position. (Maybe have a popup message that tells you to install `ex-mode` when you press <kbd>:</kbd> in command mode?)

**Comment by lloeki** [comment]
> expose motions via a service

This was, at first, the intended plan, because for now there is hard-to-maintain (or at least, harder than needed) duplication of logic and views between vim-mode and ex-mode (the whole stuff around pushops/processopstack, the "search/command-line" view and history management...).

> merge ex- and vim-mode

At the beginning I wanted things to be cleanly separated so as to be able to experiment and let the architecture emerge without impacting vim-mode, allowing to scratch an itch of my own and see if and how the community responded. This is in line with how @bhuga viewed things:

>> I made the initial decision to keep concerns separated and code clean. Smaller, focused packages are idiomatic in Atom.

As ex-mode grew, it became gradually obvious that vim- and ex- modes are indeed separate conceptually, yet bridged in a number of key areas that require a relatively rich service interface exposing some innards of vim-mode (if one wants to guarantee some API stability) or that, all the while keeping the code clean and concerns separated by internal architecture, ex-mode be integrated into vim-mode.

In the end, I feel it'd be simpler and leaner to have the vim-mode package provide what is an essential part of the vim core experience instead of having to maintain (in vim-mode) and follow (in ex-mode) a humongous API. Also, defaults matter and I'm confident people expect ex-mode to be there but don't bother downloading it for a variety of reasons.

tl;dr: for the sake of simplicity, cleaniness, and the user, I'm in favor of a merge.

**Comment by jacekkopecky** [comment]
The issue of ex `:` commands used as motions is bigger than vim-mode/ex-mode. If there's a package out there that provides extra motions (e.g. `cursor-history`, `jumpy`, and more) or when Atom provides more motions (e.g. camelCase motions), we would probably want vim-mode to integrate those seamlessly. And think about `c<click>` – change until the point where I click after pressing that `c`.

I just discovered `TextEditor::onDidChangeCursorPosition`. I hope to be able to try itegrating that as support for arbitrary motions combined with vim-mode operators.

**Comment by jacekkopecky** [comment]
Guess what, MacVIM handles `c<click>` as described above. I never knew. 8-)

**Comment by jazzpi** [comment]
It does on Linux too, probably one of the compilation options starting with `mouse` enables that.

But the other reasons of course still stand.

**Comment by maxbrunsfeld** [comment]
> Because vim treats them so differently, commands like :r !ls are a fundamentally different from a code perspective, not a memory and footprint perspective. After pressing `:`, a whole new language is parsed and evaluated.

> Smaller, focused packages are idiomatic in Atom.

I'm the current maintainer of this package, and I agree with @bhuga very strongly on this. Memory and performance-wise it's no big deal at all to add this code to `vim-mode`. But from a project-management/maintenance perspective, I really don't think it makes sense. Implementing `ex`-mode requires specific domain knowledge of an esoteric language. And we'll be able to ship improvements faster if we keep these packages separate.

That said, @jazzpi I definitely see your point that certain ex-commands *should* compose properly with vim operators and motions. I think that solving that problem is a very high-priority item on this package's roadmap.

> The issue of ex : commands used as motions is bigger than vim-mode/ex-mode. If there's a package out there that provides extra motions (e.g. cursor-history, jumpy, and more) or when Atom provides more motions (e.g. camelCase motions), we would probably want vim-mode to integrate those seamlessly.

@jacekkopecky has some great ideas about making vim-mode's operators composable with motions from other packages. I'd really like us to pursue *this* path.

@jazzpi I really apologize for not weighing in earlier on this; I know you've now put some serious work into a PR to integrate these packages. If you're interested maybe you can help in this effort to make vim-mode more modular so that `ex-mode` can integrate with it more easily.

**Comment by jazzpi** [comment]
It isn't really a problem, I knew it wasn't a guaranteed merge and honestly it was interesting reading into the codebase. I've also got a good idea about how to implement search highlighting, but no time to do so at the moment.

Two problems remain however: `ex-mode`s little exposure and the large duplicate codebase. For the exposure problem, the pop-up is the best I can come up with, does anyone else have ideas? The duplicate codebase is a bit tougher, it might be possible to expose parts via a service though.

**Comment by maxbrunsfeld** [comment]
>  For the exposure problem, the pop-up is the best I can come up with, does anyone else have ideas?

There is a proposal, (https://github.com/atom/apm/pull/385) to add a new feature to `apm` called "package sets", which will be a one-step way to install a group of related packages. `ex-mode` is one of the examples that people are using in the design of the feature. I'm not sure exactly what it will look like, but I think that when people search for `vim` on atom.io, they could be shown some kind of `vim` package set that also includes `ex-mode`.

> The duplicate codebase is a bit tougher, it might be possible to expose parts via a service though.

I agree. Ultimately, I'd like to actually *remove* the `/` command from `vim-mode`, and instead make the operators compose properly with motions caused by Atom's bundled `find-and-replace` package. For any other duplication, I'd really like to try to solve the problem by enriching the `vim-mode` service.

**Comment by dimpase** [comment]
I might be saying obvious, but ex language is something that should be trivial to tackle using modern compilation techniques, however esoteric it, as a language, might look to you.

**Comment by lee-dohm** [comment]
As stated in [the README](https://github.com/atom/vim-mode), this package is no longer maintained and is deprecated. We recommend that people use the [vim-mode-plus package](https://atom.io/packages/vim-mode-plus) instead. Because of this, we are [archiving this repository](https://help.github.com/articles/about-archiving-repositories/) and closing all issues and pull requests. Thanks very much for your support and contributions!
